File: docs/breez-sdk/snippets/rust/src/refunding_payments.rs
use anyhow::Result;
use breez_sdk_spark::*;
use log::info;

async fn list_unclaimed_deposits(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: list-unclaimed-deposits
    let request = ListUnclaimedDepositsRequest {};
    let response = sdk.list_unclaimed_deposits(request).await?;
    
    for deposit in response.deposits {
        info!("Unclaimed deposit: {}:{}", deposit.txid, deposit.vout);
        info!("Amount: {} sats", deposit.amount_sats);
        
        if let Some(claim_error) = &deposit.claim_error {
            match claim_error {
                DepositClaimError::MaxDepositClaimFeeExceeded { max_fee, required_fee, .. } => {
                    info!("Max claim fee exceeded. Max: {:?} sats, Required: {} sats", max_fee, required_fee);
                }
                DepositClaimError::MissingUtxo { .. } => {
                    info!("UTXO not found when claiming deposit");
                }
                DepositClaimError::Generic { message } => {
                    info!("Claim failed: {}", message);
                }
            }
        }
    }
    // ANCHOR_END: list-unclaimed-deposits
    Ok(())
}

async fn handle_fee_exceeded(sdk: &BreezSdk, deposit: &DepositInfo) -> Result<()> {
    // ANCHOR: handle-fee-exceeded
    if let Some(DepositClaimError::MaxDepositClaimFeeExceeded { required_fee, .. }) = &deposit.claim_error {
        // Show UI to user with the required fee and get approval
        let user_approved = true; // Replace with actual user approval logic

        if user_approved {
            let request = ClaimDepositRequest {
                txid: deposit.txid.clone(),
                vout: deposit.vout,
                max_fee: Some(Fee::Fixed { amount: *required_fee }),
            };
            sdk.claim_deposit(request).await?;
        }
    }
    // ANCHOR_END: handle-fee-exceeded
    Ok(())
}

async fn claim_deposit(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: claim-deposit
    let txid = "your_deposit_txid".to_string();
    let vout = 0;
    
    // Set a higher max fee to retry claiming
    let max_fee = Some(Fee::Fixed { amount: 5_000 });
    
    let request = ClaimDepositRequest {
        txid,
        vout,
        max_fee,
    };
    
    let response = sdk.claim_deposit(request).await?;
    info!("Deposit claimed successfully. Payment: {:?}", response.payment);
    // ANCHOR_END: claim-deposit
    Ok(())
}

async fn refund_deposit(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: refund-deposit
    let txid = "your_deposit_txid".to_string();
    let vout = 0;
    let destination_address = "bc1qexample...".to_string(); // Your Bitcoin address
    
    // Set the fee for the refund transaction using a rate
    let fee = Fee::Rate { sat_per_vbyte: 5 };
    // or using a fixed amount
    //let fee = Fee::Fixed { amount: 500 };
    
    let request = RefundDepositRequest {
        txid,
        vout,
        destination_address,
        fee,
    };
    
    let response = sdk.refund_deposit(request).await?;
    info!("Refund transaction created:");
    info!("Transaction ID: {}", response.tx_id);
    info!("Transaction hex: {}", response.tx_hex);
    // ANCHOR_END: refund-deposit
    Ok(())
}

async fn recommended_fees_example() -> Result<()> {
    // ANCHOR: recommended-fees
    let response = recommended_fees(network: Network::Mainnet).await?;
    info!("Fastest fee: {} sats/vByte", response.fastest_fee);
    info!("Half-hour fee: {} sats/vByte", response.half_hour_fee);
    info!("Hour fee: {} sats/vByte", response.hour_fee);
    info!("Economy fee: {} sats/vByte", response.economy_fee);
    info!("Minimum fee: {} sats/vByte", response.minimum_fee);
    // ANCHOR_END: recommended-fees
    Ok(())
}

File: docs/breez-sdk/snippets/go/refunding_payments.go
package example

import (
	"errors"
	"fmt"
	"log"

	"github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
)

func ListUnclaimedDeposits(sdk *breez_sdk_spark.BreezSdk) error {
	// ANCHOR: list-unclaimed-deposits
	request := breez_sdk_spark.ListUnclaimedDepositsRequest{}
	response, err := sdk.ListUnclaimedDeposits(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}

	for _, deposit := range response.Deposits {
		log.Printf("Unclaimed Deposit: %v:%v", deposit.Txid, deposit.Vout)
		log.Printf("Amount: %v sats", deposit.AmountSats)

		if claimErr := *deposit.ClaimError; claimErr != nil {
			switch claimErr := claimErr.(type) {
			case breez_sdk_spark.DepositClaimErrorMaxDepositClaimFeeExceeded:
				maxFeeStr := "none"
				if claimErr.MaxFee != nil {
					switch fee := (*claimErr.MaxFee).(type) {
					case breez_sdk_spark.FeeFixed:
						maxFeeStr = fmt.Sprintf("%v sats", fee.Amount)
					case breez_sdk_spark.FeeRate:
						maxFeeStr = fmt.Sprintf("%v sats/vByte", fee.SatPerVbyte)
					}
				}
				log.Printf("Max claim fee exceeded. Max: %v, Required: %v sats or %v sats/vByte", maxFeeStr, claimErr.RequiredFeeSats, claimErr.RequiredFeeRateSatPerVbyte)
			case breez_sdk_spark.DepositClaimErrorMissingUtxo:
				log.Print("UTXO not found when claiming deposit")
			case breez_sdk_spark.DepositClaimErrorGeneric:
				log.Printf("Claim failed: %v", claimErr.Message)
			}
		}
	}
	// ANCHOR_END: list-unclaimed-deposits
	return nil
}

func HandleFeeExceeded(sdk *breez_sdk_spark.BreezSdk, deposit breez_sdk_spark.DepositInfo) error {
	// ANCHOR: handle-fee-exceeded
	if claimErr := *deposit.ClaimError; claimErr != nil {
		if exceeded, ok := claimErr.(breez_sdk_spark.DepositClaimErrorMaxDepositClaimFeeExceeded); ok {
			requiredFee := exceeded.RequiredFeeSats

			// Show UI to user with the required fee and get approval
			userApproved := true // Replace with actual user approval logic

			if userApproved {
				maxFee := breez_sdk_spark.MaxFee(breez_sdk_spark.MaxFeeFixed{Amount: requiredFee})
				claimRequest := breez_sdk_spark.ClaimDepositRequest{
					Txid:   deposit.Txid,
					Vout:   deposit.Vout,
					MaxFee: &maxFee,
				}
				_, err := sdk.ClaimDeposit(claimRequest)
				if err != nil {
					var sdkErr *breez_sdk_spark.SdkError
					if errors.As(err, &sdkErr) {
						// Handle SdkError - can inspect specific variants if needed
						// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
					}
					return err
				}
			}
		}
	}
	// ANCHOR_END: handle-fee-exceeded
	return nil
}

func RefundDeposit(sdk *breez_sdk_spark.BreezSdk) error {
	// ANCHOR: refund-deposit
	txid := "<your_deposit_txid>"
	vout := uint32(0)
	destinationAddress := "bc1qexample..." // Your Bitcoin address

	// Set the fee for the refund transaction using the half-hour feerate
	recommendedFees, err := sdk.RecommendedFees()
	if err != nil {
		return err
	}
	fee := breez_sdk_spark.Fee(breez_sdk_spark.FeeRate{SatPerVbyte: recommendedFees.HalfHourFee})
	// or using a fixed amount
	//fee := breez_sdk_spark.Fee(breez_sdk_spark.FeeFixed{Amount: 500})
	//
	// Important: The total fee must be at least 194 sats to ensure the
	// transaction can be relayed by the Bitcoin network. If the fee is
	// lower, the refund request will be rejected.

	request := breez_sdk_spark.RefundDepositRequest{
		Txid:               txid,
		Vout:               vout,
		DestinationAddress: destinationAddress,
		Fee:                fee,
	}
	response, err := sdk.RefundDeposit(request)

	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}

	log.Print("Refund transaction created:")
	log.Printf("Transaction ID: %v", response.TxId)
	log.Printf("Transaction hex: %v", response.TxHex)
	// ANCHOR_END: refund-deposit
	return nil
}

func SetMaxFeeToRecommendedFees() error {
	// ANCHOR: set-max-fee-to-recommended-fees
	// Create the default config
	config := breez_sdk_spark.DefaultConfig(breez_sdk_spark.NetworkMainnet)
	apiKey := "<breez api key>"
	config.ApiKey = &apiKey

	// Set the maximum fee to the fastest network recommended fee at the time of claim
	// with a leeway of 1 sats/vbyte
	networkRecommendedInterface := breez_sdk_spark.MaxFee(breez_sdk_spark.MaxFeeNetworkRecommended{LeewaySatPerVbyte: 1})
	config.MaxDepositClaimFee = &networkRecommendedInterface
	// ANCHOR_END: set-max-fee-to-recommended-fees
	log.Printf("Config: %v", config)
	return nil
}

func CustomClaimLogic(sdk *breez_sdk_spark.BreezSdk, deposit breez_sdk_spark.DepositInfo) error {
	// ANCHOR: custom-claim-logic
	if claimErr := *deposit.ClaimError; claimErr != nil {
		if exceeded, ok := claimErr.(breez_sdk_spark.DepositClaimErrorMaxDepositClaimFeeExceeded); ok {
			requiredFeeRate := exceeded.RequiredFeeRateSatPerVbyte

			recommendedFees, err := sdk.RecommendedFees()
			if err != nil {
				return err
			}

			if requiredFeeRate <= recommendedFees.FastestFee {
				maxFee := breez_sdk_spark.MaxFee(breez_sdk_spark.MaxFeeRate{SatPerVbyte: requiredFeeRate})
				claimRequest := breez_sdk_spark.ClaimDepositRequest{
					Txid:   deposit.Txid,
					Vout:   deposit.Vout,
					MaxFee: &maxFee,
				}
				_, err := sdk.ClaimDeposit(claimRequest)
				if err != nil {
					var sdkErr *breez_sdk_spark.SdkError
					if errors.As(err, &sdkErr) {
						// Handle SdkError - can inspect specific variants if needed
						// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
					}
					return err
				}
			}
		}
	}
	// ANCHOR_END: custom-claim-logic
	return nil
}

func RecommendedFees(sdk *breez_sdk_spark.BreezSdk) error {
	// ANCHOR: recommended-fees
	response, err := sdk.RecommendedFees()
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError - can inspect specific variants if needed
			// e.g., switch on sdkErr variant for InsufficientFunds, NetworkError, etc.
		}
		return err
	}
	log.Printf("Fastest fee: %v sats/vByte", response.FastestFee)
	log.Printf("Half-hour fee: %v sats/vByte", response.HalfHourFee)
	log.Printf("Hour fee: %v sats/vByte", response.HourFee)
	log.Printf("Economy fee: %v sats/vByte", response.EconomyFee)
	log.Printf("Minimum fee: %v sats/vByte", response.MinimumFee)
	// ANCHOR_END: recommended-fees
	return nil
}
